functions/go/upsert-resource/upsertresource/upsert_resource.go

```diff
@@ -0,0 +1,101 @@
+package upsertresource
+
+import (
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// UpsertResource upserts resource to input list of resources
+// resources are uniquely identifies by Group, Kind, Name and Namespace
+type UpsertResource struct {
+	// Resource is the input resource for upsert
+	Resource *yaml.RNode
+}
+
+// Filter implements UpsertResource as a yaml.Filter
+func (ur UpsertResource) Filter(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	inputMeta, err := ur.Resource.GetMeta()
+	if err != nil {
+		return nodes, err
+	}
+	found := false
+	for i := range nodes {
+		rMeta, err := nodes[i].GetMeta()
+		if err != nil {
+			return nodes, err
+		}
+		// check if there is a match and replace the resource
+		if IsSameResource(inputMeta, rMeta) {
+			nodes[i], err = deepCopy(ur.Resource)
+			if err != nil {
+				return nodes, err
+			}
+			a := mergeAnnotations(inputMeta.Annotations, rMeta.Annotations)
+			err = nodes[i].SetAnnotations(a)
+			if err != nil {
+				return nodes, err
+			}
+			// found a matching resource
+			found = true
+		}
+	}
+	if !found {
+		// add resource if there is no matching resource
+		nodes = append(nodes, ur.Resource)
+	}
+	return nodes, nil
+}
+
+// IsSameResource returns true if metadata of two resources
+// have same Group, Kind, Name, Namespace
+func IsSameResource(meta1, meta2 yaml.ResourceMeta) bool {
```

Comment:
This is a fundamental concept that should eventually be in the library. For example, see this function in the TS SDK: https://googlecontainertools.github.io/kpt-functions-sdk/api/modules/_types_.html#kuberneteskey

Reply:
@mengqiy Another thing to address in post 1.0 focus on function SDK

Reply:
Ack

Reply:
How are you tracking these changes that need to be made to the SDK? Issue? TODO here?

Reply:
Added TODO and created an issue. https://github.com/GoogleContainerTools/kpt/issues/2043